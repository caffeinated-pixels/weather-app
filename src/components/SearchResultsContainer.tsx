import { forwardRef, ForwardedRef, ReactNode } from 'react'
import { CloseResults } from './SearchResults'
import { FetchLocationResults } from '../WeatherDataContext'

type SearchResultsContainerProps = {
  closeResults: CloseResults
  headerText: string
  fetchLocationResults: FetchLocationResults
  children: ReactNode
}

export const SearchResultsContainer = forwardRef(
  (
    {
      closeResults,
      headerText,
      fetchLocationResults,
      children,
    }: SearchResultsContainerProps,
    ref: ForwardedRef<HTMLDivElement>
  ) => {
    return (
      <div
        id="search-results"
        className="search-result-wrapper"
        tabIndex={0}
        ref={ref}
        onBlur={closeResults}
      >
        <div className="search-result-header">
          <p>{headerText}</p>
          <button
            className="icon-button"
            onClick={() => fetchLocationResults()}
            aria-label="close search results"
          >
            <i className="fas fa-times"></i>
          </button>
        </div>
        {children}
      </div>
    )
  }
)
