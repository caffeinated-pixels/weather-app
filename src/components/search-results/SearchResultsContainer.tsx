import { ForwardedRef, forwardRef, ReactNode } from 'react'

import { FetchLocationResults } from '../../WeatherDataContext'
import { CloseResults } from './SearchResults'

type SearchResultsContainerProps = {
  closeResults: CloseResults
  headerText: string
  fetchLocationResults: FetchLocationResults
  children: ReactNode
}

export const SearchResultsContainer = forwardRef(
  (
    {
      closeResults,
      headerText,
      fetchLocationResults,
      children,
    }: SearchResultsContainerProps,
    ref: ForwardedRef<HTMLDivElement>
  ) => {
    return (
      <div
        id="search-results"
        className="search-result-wrapper"
        tabIndex={0}
        ref={ref}
        onBlur={closeResults}
      >
        <div className="search-result-header">
          <p>{headerText}</p>
          <button
            className="icon-button"
            onClick={() => fetchLocationResults()}
            aria-label="close search results"
          >
            <i className="fas fa-times"></i>
          </button>
        </div>
        {children}
      </div>
    )
  }
)
