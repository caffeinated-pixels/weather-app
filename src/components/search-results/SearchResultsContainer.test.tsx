import { render, screen, fireEvent } from '@testing-library/react'
import { SearchResultsContainer } from './SearchResultsContainer'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'

const closeResults = jest.fn()
const headerText = 'Search Results'
const fetchLocationResults = jest.fn()
const children = <div>Results</div>

const renderSearchResultsContainer = () => {
  return render(
    <SearchResultsContainer
      closeResults={closeResults}
      headerText={headerText}
      fetchLocationResults={fetchLocationResults}
    >
      {children}
    </SearchResultsContainer>
  )
}

describe('SearchResultsContainer', () => {
  it('should match the snapshot', () => {
    const domTree = renderer
      .create(
        <SearchResultsContainer
          closeResults={closeResults}
          headerText={headerText}
          fetchLocationResults={fetchLocationResults}
        >
          {children}
        </SearchResultsContainer>
      )
      .toJSON()
    expect(domTree).toMatchSnapshot()
  })

  it('renders the header text', () => {
    renderSearchResultsContainer()

    expect(screen.getByText(headerText)).toBeInTheDocument()
  })

  it('renders the children', () => {
    renderSearchResultsContainer()

    expect(screen.getByText('Results')).toBeInTheDocument()
  })

  it('calls the closeResults function when blurred', () => {
    renderSearchResultsContainer()

    fireEvent.blur(screen.getByRole('button'))

    expect(closeResults).toHaveBeenCalled()
  })

  it('calls the fetchLocationResults function when the close button is clicked', async () => {
    renderSearchResultsContainer()

    await userEvent.click(screen.getByLabelText('close search results'))

    expect(fetchLocationResults).toHaveBeenCalled()
  })
})
