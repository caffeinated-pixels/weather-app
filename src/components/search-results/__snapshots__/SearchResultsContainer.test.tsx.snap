// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsContainer should match the snapshot 1`] = `
<div
  className="search-result-wrapper"
  id="search-results"
  onBlur={[MockFunction]}
  tabIndex={0}
>
  <div
    className="search-result-header"
  >
    <p>
      Search Results
    </p>
    <button
      aria-label="close search results"
      className="icon-button"
      onClick={[Function]}
    >
      <i
        className="fas fa-times"
      />
    </button>
  </div>
  <div>
    Results
  </div>
</div>
`;
